### PR TITLE
Fix cumprod zero gradient 119430

### LINKED ISSUE #119430: `tf.math.cumprod` gradient is incorrect when input contains zeros
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

Yes

### Source

source

### TensorFlow version

 TensorFlow: 2.13.1 (also reproducible on 2.16+)

### Custom code

Yes

### OS platform and distribution

_No response_

### Mobile device

_No response_

### Python version

_No response_

### Bazel version

_No response_

### GCC/compiler version

_No response_

### CUDA/cuDNN version

_No response_

### GPU model and memory

_No response_

### Summary

`tf.math.cumprod`'s gradient computation produces wrong values when any element of the input tensor is zero. The gradient for the zero element (and sometimes adjacent elements) is incorrectly returned as `0` instead of the mathematically correct value. This affects both eager mode and XLA.

### Root cause

The gradient implementation likely computes `d/dx[i]` using a formula involving `cumprod(x) / x[i]`. When `x[i] = 0`, this produces `0/0` because `cumprod[j]` for all `j >= i` is also `0`. The `0/0` evaluates to `NaN`, which is then silently replaced with `0`.

The correct approach is to compute the gradient using exclusive cumulative products from both directions (prefix and suffix), avoiding division entirely.

### Additional affected cases

| Input | TF gradient | Correct gradient | Wrong elements |
|-------|-------------|-----------------|----------------|
| `[0, 2, 3]` | `[0, 0, 0]` | `[9, 0, 0]` | `x[0]` |
| `[2, 3, 0]` | `[4, 2, 0]` | `[4, 2, 6]` | `x[2]` |
| `[2, 0, 0, 3]` | `[1, 0, 0, 0]` | `[1, 2, 0, 0]` | `x[1]` |

Note: when no zeros are present, the gradient is correct.

### Impact

Any training loop that uses `cumprod` on data that can contain zeros will silently receive wrong gradients, potentially causing the model to fail to learn or learn incorrectly. This includes survival analysis models (hazard functions), attention mechanisms with zero-masking, and probabilistic sequence models.

### Environment

- TensorFlow: 2.13.1 (also reproducible on 2.16+)
- OS: Ubuntu 20.04
- CPU-only (`CUDA_VISIBLE_DEVICES=""`)
- Also affects XLA (`jit_compile=True`)

### Standalone code to reproduce the issue

```shell
### Reproduction


import os
os.environ["CUDA_VISIBLE_DEVICES"] = ""
import tensorflow as tf
import numpy as np

x = tf.constant([1.5, 0.0, 2.0, 3.0], dtype=tf.float64)
with tf.GradientTape() as tape:
    tape.watch(x)
    y = tf.reduce_sum(tf.math.cumprod(x))
g = tape.gradient(y, x)
print("TF gradient:", g.numpy())       # [1. 0. 0. 0.]

# Numerical verification (central difference)
def f(xv):
    return float(np.sum(np.cumprod(xv)))
xv = np.array([1.5, 0.0, 2.0, 3.0])
num_g = np.zeros(4)
for i in range(4):
    xp, xm = xv.copy(), xv.copy()
    xp[i] += 1e-7; xm[i] -= 1e-7
    num_g[i] = (f(xp) - f(xm)) / 2e-7
print("Correct gradient:", num_g)       # [1. 13.5 0. 0.]


**Output:**

TF gradient: [1. 0. 0. 0.]
Correct gradient: [ 1.  13.5  0.   0. ]


The gradient for `x[1]` should be `13.5`, not `0`.

### Mathematical derivation

For `cumprod([x₀, x₁, x₂, x₃]) = [x₀, x₀x₁, x₀x₁x₂, x₀x₁x₂x₃]`:


d/dx₁ sum(cumprod) = d/dx₁ (x₀ + x₀x₁ + x₀x₁x₂ + x₀x₁x₂x₃)
                   = x₀ + x₀x₂ + x₀x₂x₃
                   = 1.5 + 1.5*2 + 1.5*2*3
                   = 1.5 + 3.0 + 9.0
                   = 13.5
```

### Relevant log output

```shell

```

### PR DESCRIPTION
**Pull request by danyalahmed1995**
Fixes #119430.

`Cumprod`'s gradient currently uses the existing division-based path for all elements. That works for non-zero inputs, but drops valid gradient contributions when `x == 0`.

This keeps the existing path for non-zero elements and adds zero-specific handling that only contributes terms whose cumulative product contains exactly one zero.

Added regression coverage for zero positions at the beginning, middle, and end, multiple zeros, `axis=1`, `exclusive`, `reverse`, and `exclusive + reverse`.

Validation:

* `python -m py_compile tensorflow/python/ops/math_grad.py tensorflow/python/ops/math_grad_test.py`

I also attempted the focused Bazel target in fork CI, but TensorFlow source compilation exceeded the 90-minute Actions timeout before the test executed.

### CODE REVIEW

**Review by dmiltr3 — CHANGES_REQUESTED** [review]
Hi @danyalahmed1995,
  
  Thank you for this excellent pull request! We reviewed the mathematical implementation of the changes and confirmed everything is correct:
  
  1. **Mathematical Correctness**: The zero-specific handling correctly resolves the `NaN` / zero-gradient drop from the division-based path.
  2. **Performance**: Excellent work keeping the time and space complexity to $O(N)$ without nesting or generating $O(N^2)$ intermediate tensors. This will scale well on long sequence scans. 
  3. **XLA Compatibility**: Verified that the new logic compiles and executes correctly under eager mode, graph mode, and XLA backends (CPU/GPU/TPU).
  
  We would love to merge this! To finalize the PR, please add the following comprehensive unit tests to `tensorflow/python/ops/math_grad_test.py` to ensure regression coverage for all zero-handling combinations:
  
  ```python
  @test_util.run_all_in_graph_and_eager_modes
  class CumprodGradTest(test.TestCase):
  
    def _test_cumprod_grad(self, x_in, axis=0, exclusive=False, reverse=False):
      x = constant_op.constant(x_in, dtype=dtypes.float64)
      with backprop.GradientTape() as tape:
        tape.watch(x)
        y = math_ops.cumprod(x, axis=axis, exclusive=exclusive, reverse=reverse)
        loss = math_ops.reduce_sum(y)
      grad = tape.gradient(loss, x)
      return self.evaluate(grad)
  
    def testZeroPositions(self):
      # Zero at beginning
      g = self._test_cumprod_grad([0.0, 2.0, 3.0])
      self.assertAllClose(g, [9.0, 0.0, 0.0])
  
      # Zero in middle
      g = self._test_cumprod_grad([2.0, 0.0, 3.0])
      self.assertAllClose(g, [1.0, 8.0, 0.0])
  
      # Zero at end
      g = self._test_cumprod_grad([2.0, 3.0, 0.0])
      self.assertAllClose(g, [4.0, 2.0, 6.0])
  
      # Multiple zeros
      g = self._test_cumprod_grad([2.0, 0.0, 3.0, 0.0])
      self.assertAllClose(g, [1.0, 8.0, 0.0, 0.0])
  
    def testZeroPositionsExclusive(self):
      # Zero at beginning
      g = self._test_cumprod_grad([0.0, 2.0, 3.0], exclusive=True)
      self.assertAllClose(g, [3.0, 0.0, 0.0])
  
      # Zero in middle
      g = self._test_cumprod_grad([2.0, 0.0, 3.0], exclusive=True)
      self.assertAllClose(g, [1.0, 2.0, 0.0])
  
      # Zero at end
      g = self._test_cumprod_grad([2.0, 3.0, 0.0], exclusive=True)
      self.assertAllClose(g, [4.0, 2.0, 0.0])
  
      # Multiple zeros
      g = self._test_cumprod_grad([2.0, 0.0, 3.0, 0.0], exclusive=True)
      self.assertAllClose(g, [1.0, 2.0, 0.0, 0.0])
  
    def testZeroPositionsReverse(self):
      # Zero at beginning
      g = self._test_cumprod_grad([0.0, 2.0, 3.0], reverse=True)
      self.assertAllClose(g, [6.0, 3.0, 3.0])
  
      # Zero in middle
      g = self._test_cumprod_grad([2.0, 0.0, 3.0], reverse=True)
      self.assertAllClose(g, [0.0, 9.0, 1.0])
  
      # Zero at end
      g = self._test_cumprod_grad([2.0, 3.0, 0.0], reverse=True)
      self.assertAllClose(g, [0.0, 0.0, 10.0])
  
      # Multiple zeros
      g = self._test_cumprod_grad([2.0, 0.0, 3.0, 0.0], reverse=True)
      self.assertAllClose(g, [0.0, 0.0, 0.0, 4.0])
  
    def testZeroPositionsExclusiveReverse(self):
      # Zero at beginning
      g = self._test_cumprod_grad([0.0, 2.0, 3.0], exclusive=True, reverse=True)
      self.assertAllClose(g, [0.0, 3.0, 3.0])
  
      # Zero in middle
      g = self._test_cumprod_grad([2.0, 0.0, 3.0], exclusive=True, reverse=True)
      self.assertAllClose(g, [0.0, 3.0, 1.0])
  
      # Zero at end
      g = self._test_cumprod_grad([2.0, 3.0, 0.0], exclusive=True, reverse=True)
      self.assertAllClose(g, [0.0, 0.0, 3.0])
  
      # Multiple zeros
      g = self._test_cumprod_grad([2.0, 0.0, 3.0, 0.0], exclusive=True, reverse=True)
      self.assertAllClose(g, [0.0, 0.0, 0.0, 4.0])
  
    def testAxis1(self):
      g = self._test_cumprod_grad([[2.0, 0.0], [3.0, 4.0]], axis=1)
      self.assertAllClose(g, [[1.0, 2.0], [5.0, 3.0]])
  
  ```
Once these changes are added, please let us know so we can complete the review process and merge the PR!

**Comment by danyalahmed1995** [comment]
Hello @dmiltr3 thank you for the response , I didn’t apply the suggested block literally because this PR already has a `CumprodGradTest` class and `_cumprod_grad` helper.

Adding it as it is would duplicate the existing coverage and redefine the test class, which would shadow the original class during discovery.

I added the requested zero-position matrix to the existing class instead, using the existing helper and keeping the current dtype and axis coverage intact.

But if you prefer your original suggestion to be added , kindly let me know .

**Review by dmiltr3 — APPROVED** [review]

**Comment by dmiltr3** [comment]
The implementation is clean, runs in $O(N)$ time and space complexity, works properly with XLA, and includes robust test coverage.
